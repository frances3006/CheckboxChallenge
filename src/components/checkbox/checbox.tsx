
import * as React from "react";
import * as styles from './checkbox.css';
import * as cn from 'classnames';

interface IProps { 
    labelText: string;
    disabled?: boolean;
    onChange?: (label: string) => any;
};

interface IState {
    checkStyle: string;
    clicked: boolean
};

export class Checkbox extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);
        this.state = {
            checkStyle: styles.checkbox,
            clicked: false
        }
    }

    onClick = (event: any) => {
        if(!this.props.disabled) {
          this.setState({
            checkStyle: this.state.clicked ? styles.checkbox : cn(styles.checkbox, styles.checkboxChecked),
            clicked: !this.state.clicked
          });
          if(this.props.onChange) {
            this.props.onChange(this.props.labelText);
          }
        }
      }
    
      mouseEnter = (event: any) => {
        if(!this.props.disabled) {
            this.setState({
                checkStyle: cn(styles.checkbox, styles.checkboxEnter)
              })
        }    
      }
    
      mouseLeave = (event: any) => {
        this.setState({
            checkStyle: this.state.clicked ? cn(styles.checkbox, styles.checkboxChecked) : styles.checkbox
        })
      }
    
      render() {
        const { checkStyle } = this.state;
        const checkboxClass = this.props.disabled ? 
          cn(checkStyle, styles.checkboxDisabled ) : 
          cn(checkStyle);

        const wrapperClass = this.props.disabled ? 
          cn(styles.wrapper, styles.wrapperDisabled) : styles.wrapper; 
        return (
          <label 
            className={wrapperClass}
            onClick={this.onClick} 
            onMouseEnter={this.mouseEnter}
            onMouseLeave={this.mouseLeave}>
              <div className={checkboxClass}/>
              {this.props.labelText}
          </label>
            
        );
      }

}