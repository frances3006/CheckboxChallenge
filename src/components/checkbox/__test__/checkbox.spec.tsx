import * as React from "react";
import { shallow } from "enzyme";

import { Checkbox } from '../checbox';

test('checkbox snapshot', () => {
    const shallowWrapper = shallow(
        <Checkbox labelText='Hello'/>
    );
    expect(shallowWrapper).toMatchSnapshot();
});

test('checkbox onClick', () => {
    const onChange = jest.fn();

    const shallowWrapper = shallow(
        <Checkbox labelText='Hello' onChange={onChange}/>
    );
    shallowWrapper.find('.wrapper').simulate('click');

    expect(onChange).toBeCalledWith('Hello');
});

test('checkbox disabled', () => {
    const onChange = jest.fn();

    const shallowWrapper = shallow(
        <Checkbox labelText='Hello' disabled={true} onChange={onChange}/>
    );
    shallowWrapper.find('.wrapper').simulate('click');

    expect(onChange).toHaveBeenCalledTimes(0);
});

test('checkbox changes css when checked', () => {
    const shallowWrapper = shallow(
        <Checkbox labelText='Hello'/>
    );

    expect(shallowWrapper.find('.checkboxUnchecked')).toHaveLength(1);
    shallowWrapper.find('.wrapper').simulate('click');
    expect(shallowWrapper.find('.checkboxUnchecked')).toHaveLength(0);
    expect(shallowWrapper.find('.checkboxChecked')).toHaveLength(1);
});

test('checkbox css not changing when disabed', () => {
    const shallowWrapper = shallow(
        <Checkbox labelText='Hello' disabled={true}/>
    );

    expect(shallowWrapper.find('.checkboxUnchecked')).toHaveLength(1);
    shallowWrapper.find('.wrapper').simulate('click');
    expect(shallowWrapper.find('.checkboxUnchecked')).toHaveLength(1);
    expect(shallowWrapper.find('.checkboxChecked')).toHaveLength(0);
});