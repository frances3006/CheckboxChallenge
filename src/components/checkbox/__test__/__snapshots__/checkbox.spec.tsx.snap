// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Checkbox
    labelText="Hello"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="checkboxUnchecked"
        />,
        "Hello",
      ],
      "className": "wrapper",
      "onClick": [Function],
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "checkboxUnchecked",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "Hello",
    ],
    "type": "label",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="checkboxUnchecked"
          />,
          "Hello",
        ],
        "className": "wrapper",
        "onClick": [Function],
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "checkboxUnchecked",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "Hello",
      ],
      "type": "label",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
